File: UML/composition.py
## strong "has-a" relationship
## One object completely owns another object.
# If the parent object is destroyed, the child object is also destroyed.

class Engine:
    def start(self):
        return "Engine started"


class Car:
    def __init__(self):
        self.engine = Engine()   # created inside → composition
    
    def start_car(self):
        return self.engine.start()

c1 = Car()
print(c1.start_car())

"""
class Car:
    def __init__(self, engine):
        self.engine = engine   # passed from outside
"""

"""
class Car:
    def __init__(self, engine):
        self.engine = engine   # passed from outside
"""

File: UML/composition.go
// composition in golang

package UML

import "fmt"

type Engine struct {
    Power int
}

func (e Engine) Start() {
    fmt.Println("Engine started")
}

type Car struct {
    Engine
    Brand string
}

func main() {
    c := Car{Engine: Engine{Power: 150}, Brand: "Toyota"}
    c.Start() 
}